framework/web/MultipartFormDataParser.php

```diff
@@ -0,0 +1,357 @@
+ * means [[UploadedFile::saveAs()]] will fail as well.
+ *
+ * @property integer $uploadFileMaxSize upload file max size in bytes.
+ * @property integer $uploadFileMaxCount maximum upload files count.
+ *
+ * @author Paul Klimov <klimov.paul@gmail.com>
+ * @since 2.0.10
+ */
+class MultipartFormDataParser extends Object implements RequestParserInterface
+{
+    /**
+     * @var integer upload file max size in bytes.
+     */
+    private $_uploadFileMaxSize;
+    /**
+     * @var integer maximum upload files count.
+     */
+    private $_uploadFileMaxCount;
+
+
+    /**
+     * @return integer upload file max size in bytes.
+     */
+    public function getUploadFileMaxSize()
+    {
+        if ($this->_uploadFileMaxSize === null) {
+            $this->_uploadFileMaxSize = $this->getByteSize(ini_get('upload_max_filesize'));
+        }
+        return $this->_uploadFileMaxSize;
+    }
+
+    /**
+     * @param integer $uploadFileMaxSize upload file max size in bytes.
+     */
+    public function setUploadFileMaxSize($uploadFileMaxSize)
+    {
+        $this->_uploadFileMaxSize = $uploadFileMaxSize;
+    }
+
+    /**
+     * @return integer maximum upload files count.
+     */
+    public function getUploadFileMaxCount()
+    {
+        if ($this->_uploadFileMaxCount === null) {
+            $this->_uploadFileMaxCount = ini_get('max_file_uploads');
+        }
+        return $this->_uploadFileMaxCount;
+    }
+
+    /**
+     * @param integer $uploadFileMaxCount maximum upload files count.
+     */
+    public function setUploadFileMaxCount($uploadFileMaxCount)
+    {
+        $this->_uploadFileMaxCount = $uploadFileMaxCount;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function parse($rawBody, $contentType)
+    {
+        if (!empty($_POST) || !empty($_FILES)) {
+            // normal POST request is parsed by PHP automatically
+            return $_POST;
+        }
+
+        if (empty($rawBody)) {
+            return [];
+        }
+
+        if (!preg_match('/boundary=(.*)$/is', $contentType, $matches)) {
+            return [];
+        }
+        $boundary = $matches[1];
+
+        $bodyParts = preg_split('/-+' . preg_quote($boundary) . '/s', $rawBody);
+        array_pop($bodyParts); // last block always has no data
+
+        $bodyParams = [];
+        $filesCount = 0;
+        foreach ($bodyParts as $bodyPart) {
+            if (empty($bodyPart)) {
+                continue;
+            }
+            list($headers, $value) = explode("\r\n\r\n", $bodyPart, 2);
+            $headers = $this->parseHeaders($headers);
+            
+            if (!isset($headers['content-disposition']['name'])) {
+                continue;
+            }
+
+            if (isset($headers['content-disposition']['filename'])) {
+                // file upload:
+                if ($filesCount >= $this->getUploadFileMaxCount()) {
+                    continue;
```

Comment:
Currently it's on par with PHP behavior but I agree that exception is a bit more developer-friendly.